common: added a field in HCM proto to be able to reverse the order of HTTP encoder filters.

qiannawang opened the pull request:
*Description*: Added a field in HCM proto to be able to reverse the order of HTTP encoder filters. The field is set false by default, indicating HTTP encoder filters have the same order as configured in the filter
chain. If true, their order will be reversed.
*Risk Level*: low
*Testing*: bazel test //test/...
Part of #4599

Signed-off-by: Qi (Anna) Wang <qiwang@google.com>

qiannawang commented:
@htuch could you PTAL?

rshriram commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -348,6 +348,10 @@ message HttpConnectionManager {
+
+  // If true, the order of encoder filters will be reversed to that of filters
+  // configured in the HTTP filter chain.
+  bool reverse_encoders_order = 27;
```
what is the use case? it would be good to have some explanation associated with such settings. Otherwise, we end up having a plethora of "invert_x", "reverse_y" boolean options all over the config

qiannawang replied:
This is a bug fixing in envoy. The design of envoy was to have reserved order of encoder filters to the configuration in the filter chain. However, the current implementation in envoy is not following this order. In order to avoid breaking anyone who depends on this bug, we decided to fix this as an opt-in behavior.

For more details, see https://github.com/envoyproxy/envoy/issues/4599

dnoe replied:
Does it make sense to link https://github.com/envoyproxy/envoy/issues/4599 in the comment for this? It adds valuable context.

rshriram replied:
I don't think we have to go to this extreme to preserve backward compatibility for incorrect behavior. It seems counter intuitive, given that the goal of the API is to describe the system (not workarounds for "correct" behavior). Put another way, if those other systems depended on the incorrect behavior, then those systems are also not functioning correctly are they?

qiannawang replied:
As @htuch suggested, we will mark this field deprecated and put in the release note too.

dnoe commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -199,8 +199,14 @@ class HttpConnectionManagerImplTest : public Test, public ConnectionManagerConfi
-    for (auto filter : encoder_filters_) {
-      EXPECT_CALL(*filter, onDestroy());
+    if (reverse_encoders_order_) {
+      for (int i = encoder_filters_.size() - 1; i >= 0; --i) {
```
This might be clearer using `std::for_each` with reversed and regular iterators. Then you can do something like:

```
auto setup_filter_expect = [](MockStreamEncoderFilter* filter){ EXPECT_CALL(*filter, onDestroy()); };
if (reverse_encoders_order_) {
  std::for_each(encoder_filters.rbegin(), encoder_filters.rend(), setup_filter_expect);
} else {
  std::for_each(encoder_filters.begin(), encoder_filters.end(), setup_filter_expect);
```

This should let you write the same style code for reverse and forward, which seems to have some readability benefit.

qiannawang replied:
@dnoe good idea! Incorporated your suggestion.

rshriram commented:
At the very least, if we are going to do this "bugfix", might as well come clean and declare the field as "bugfix_reverse_encoder_order", so that end users don't mistake this for a "new" feature. 

I still think providing backward compatibility for a buggy behavior is a strange thing to do

lizan commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -348,6 +348,10 @@ message HttpConnectionManager {
+
+  // If true, the order of encoder filters will be reversed to that of filters
+  // configured in the HTTP filter chain.
+  bool reverse_encoders_order = 27;
```
Should we make this BoolValue and change the default to true in near future (after next release)?

qiannawang replied:
Done with the renaming.

@htuch would you recommend to add the "deprecated" annotation in this change or in another PR when we deprecate the field in that release?

 @lizan what would be the difference between bool and BoolValue in this case?

lizan replied:
bool is always default to false and you can't differentiate false and unset. with BoolValue you know if it is unset or true or false.

lizan replied:
The question is what will be the process when we remove this deprecated field. If at that moment this will be true (true is expected behavior, right?) then BoolValue makes more sense. @htuch?

qiannawang replied:
hmm, the h1_capture_direct_response_fuzz_test test is hitting tsan failure after I set the bool filed [depreated = true]. I am wondering if the BoolValue would help with the failure.

https://circleci.com/gh/envoyproxy/envoy/108918?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

lizan replied:
It is a flaky one and fix in recent commit, merge master would help.

qiannawang replied:
Merging does help, but looks like I rebased it incorrectly and the version history is corrupted :(

Any suggestions to fix this? I guess I can close this PR and create a new one, but wonder if there is a simple solution.

lizan replied:
Yes you shouldn't do `git rebase` but a simple `git merge upstream/master`. you can checkout the original commit and do git merge again, and then force push to fix this PR, or close and create a new PR.

qiannawang replied:
It works! thanks for the suggestion!

qiannawang commented on source/common/http/conn_manager_config.h:
```diff
@@ -209,6 +209,12 @@ class ConnectionManagerConfig {
   virtual FilterChainFactory& filterFactory() PURE;
 
+  /**
+   * @return whether the connection manager will reverse the order of encoder
```
Ack.

qiannawang commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -458,7 +458,15 @@ void ConnectionManagerImpl::ActiveStream::addStreamEncoderFilterWorker(
+  if (connection_manager_.config_.reverseEncodersOrder()) {
+    wrapper->moveIntoList(std::move(wrapper), encoder_filters_);
+  } else {
+    if (connection_manager_.config_.reverseEncodersOrder()) {
```
Removed.

qiannawang commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -199,8 +199,14 @@ class HttpConnectionManagerImplTest : public Test, public ConnectionManagerConfi
-    for (auto filter : encoder_filters_) {
-      EXPECT_CALL(*filter, onDestroy());
+    if (reverse_encoders_order_) {
+      for (int i = encoder_filters_.size() - 1; i >= 0; --i) {
```
Yes, as the InSequence is global within a test case for all callers.

qiannawang commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -3206,6 +3246,95 @@ TEST_F(HttpConnectionManagerImplTest, MultipleFilters) {
   EXPECT_EQ(ssl_connection_.get(), encoder_filters_[1]->callbacks_->connection()->ssl());
 }
 
+TEST_F(HttpConnectionManagerImplTest, MultipleFiltersWithReversedEncoders) {
```
Done.

htuch commented:
Agree with @rshriram that this should be named as a bug fix. I think it's arguably better to support the old behavior here as we can break folks in production who have unexpected encode filter ordering interactions.

In terms of deprecation, let's immediately deprecate the field and update DEPRECATED.md.

htuch commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -3206,6 +3245,97 @@ TEST_F(HttpConnectionManagerImplTest, MultipleFilters) {
   EXPECT_EQ(ssl_connection_.get(), encoder_filters_[1]->callbacks_->connection()->ssl());
 }
 
+// MultipleFiltersWithReversedEncoders verifies the same case of MultipleFilters, except that the
```
Are these the only tests that depend on order? If there are others, I think we should convert them to the reverse order and only keep around one or two tests for the existing order.

qiannawang replied:
As per off-line discussion, set the default value to true in the test and left two tests (with suffix of .*InorderEncoders) verifying the existing order.

htuch commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -348,6 +348,10 @@ message HttpConnectionManager {
+
+  // If true, the order of encoder filters will be reversed to that of filters
+  // configured in the HTTP filter chain.
+  bool reverse_encode_order = 27;
```
See other comment on naming/deprecation.

qiannawang replied:
Done.

htuch commented:
@lizan we are beginning immediate deprecation. Can you explain how defaulting the field to true at the end of the deprecation cycle, at the same time as we remove it, makes more sense? I don't mind either way TBH, just trying to grok the right thing to do here.

lizan commented:
@htuch I don't mind either, if we do default to true and remove at same time, then no difference, if we do remove later there will be less surprise.

@qiannawang looks like you reverted too much?

qiannawang commented:
> @htuch I don't mind either, if we do default to true and remove at same time, then no difference, if we do remove later there will be less surprise.
> 
> @qiannawang looks like you reverted too much? I don't see the deprecation now.

I have added back the deprecation, resolved all comments and changed to use BoolValue from bool. 

@htuch PTAL?

qiannawang commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -348,6 +348,12 @@ message HttpConnectionManager {
   };
   repeated UpgradeConfig upgrade_configs = 23;
+
+  // If true, the order of encoder filters will be reversed to that of filters
```
Make sense to me. Updated with a more descriptive comment in the proto.

 I noticed ci/circleci: mac test (not required) is failing. I do not see the failure related to my change. Is this a concern?

lizan replied:
Thanks. re CI: Looks like it is an intermittent error from CircleCI, should be ok.

htuch commented on DEPRECATED.md:
```diff
@@ -8,6 +8,8 @@ A logged warning is expected for each deprecated item that is in deprecation win
 
 ## Version 1.9.0 (pending)
 
+* Use of the boolean field `bugfix_reverse_encode_order` in [http_connection_manager.proto](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto) is deprecated. This is a bug fix to resolve issue [#4599](https://github.com/envoyproxy/envoy/issues/4599) for envoy where it is designed to have the encode filters in r ...
```
Can you point out at the beginning of this bullet point something to the effect of:

"Order of execution of the encoder filter chain has been reversed. Prior to this release cycle it was incorrect, see `#4599`. In the 1.9.0 release cycle we introduced `bugfix_reverse_encode_order` to temporarily support both old and new behaviors."

Then the rest can be pretty much what you have there..

qiannawang replied:
Done.

htuch commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -348,6 +348,14 @@ message HttpConnectionManager {
+  // If true, the order of encoder filters will be reversed to that of filters
+  // configured in the HTTP filter chain. Otherwise, it will keep the existing
+  // order.
+  // Note: this is a bug fix for envoy, which is designed to have the reversed
```
s/envoy/Envoy/, can you reference the GH issue here as well?

qiannawang replied:
Done.

htuch commented on source/common/config/filter_json.cc:
```diff
@@ -209,6 +209,7 @@ void FilterJson::translateHttpConnectionManager(
 
   JSON_UTIL_SET_BOOL(json_config, proto_config, use_remote_address);
   JSON_UTIL_SET_BOOL(json_config, proto_config, generate_request_id);
+  JSON_UTIL_SET_BOOL(json_config, proto_config, bugfix_reverse_encode_order);
```
Why is this needed? This is just for v1 config, which is now deprecated.

qiannawang replied:
Removed.

qiannawang commented:
All tests pass. @htuch PTAL?
